k2/csrc/cuda/array.h

```diff
@@ -403,19 +407,22 @@ class Array2 {
     region_ = a.GetRegion();
   }
 
-  // Warning: user should never call this function, we declare
+  // Caution: user should never call this function, we declare
   // it as public just because the enclosing parent function for an
   // extended __host__ __device__ lambda
   // must have public access.
   void CopyDataFromTensor(const Tensor &t) {
     T *this_data = Data();
     const T *t_data = t.Data<T>();
     int32_t elem_stride1 = t.GetShape().Stride(1);
-    auto lambda_copy_elems = [=] __host__ __device__(int32_t i, int j)->void {
-      this_data[i * elem_stride0_ + j] =
-          t_data[i * elem_stride0_ + j * elem_stride1];
+    auto lambda_copy_elems = [=] __host__ __device__(
+                                     int32_t i, int32_t j, int32_t elem_stride0,
+                                     int32_t elem_stride1)->void {
+      this_data[i * elem_stride0 + j] =
+          t_data[i * elem_stride0 + j * elem_stride1];
     };
-    Eval2(region_->context, dim0_, dim1_, lambda_copy_elems);
+    Eval2(region_->context, dim0_, dim1_, lambda_copy_elems, elem_stride0_,
```

Comment:
Sometims there are additional parameters  required by `lambda` but they are on CPU memory (e.g. `elem_stride0_` and `elem_stride1` here), we need a way to convert them to GPU memory automatically. Here I use `parameter pack` to accept and forward those parameters. See my updates for `Eval2` below for details.